src/middlewared/middlewared/plugins/api_key.py

```diff
@@ -22,161 +27,356 @@ class APIKeyModel(sa.Model):
+                    # Feed SID we looked up back into our extend context
+                    # Because there may be multiple keys for same SID value
+                    ctx['by_sid'][user_identifier] = username
+
+            if username:
+                item['username'] = username
+        else:
+            # Something wildly invalid got written, but we can't
+            # write a log message here (queried too frequently).
+            item['username'] = None
+            item['local'] = True
+
+        if item['username'] is None:
+            # prevent keys we can't resolve from being written
+            item['revoked'] = True
+
+        match expiry:
+            case -1:
+                # key has been forcibly revoked
+                item['revoked'] = True
+            case 0:
+                # zero value indicates never expires
+                pass
+            case _:
+                item['expires_at'] = datetime.fromtimestamp(expiry, UTC)
+
         return item
 
-    @api_method(ApiKeyCreateArgs, ApiKeyCreateResult, audit='Create API key', audit_extended=lambda data: data['name'])
-    async def do_create(self, data: dict) -> dict:
+    @private
+    def compress(self, data: dict) -> dict:
+        out = data.copy()
+        if 'expires_at' in out:
+            if (expires_at := out.pop('expires_at')) is None:
+                out['expiry'] = 0
+            else:
+                out['expiry'] = int(expires_at.timestamp())
+
+        if out.get('revoked'):
+            out['expiry'] = -1
+
+        thehash = out.pop('keyhash')
+        if thehash:
+            out['key'] = thehash
+
+        for key in [
+            'username',
+            'revoked',
+            'keyhash',
+            'local',
+        ]:
+            out.pop(key, None)
+
+        return out
+
+    @api_method(
+        ApiKeyCreateArgs,
+        ApiKeyCreateResult,
+        audit='Create API key',
+        audit_extended=lambda data: data['name'],
+        roles=['READONLY_ADMIN', 'API_KEY_WRITE']
+    )
+    @pass_app(rest=True)
+    def do_create(self, app, data: dict) -> dict:
         """
         Creates API Key.
 
         `name` is a user-readable name for key.
         """
-        await self._validate("api_key_create", data)
-
-        key = self._generate()
-        data["key"] = pbkdf2_sha256.encrypt(key)
+        # First catch any privilege errors to avoid leaking potentially sensitive information
+        self.api_key_privilege_check(app, data['username'], 'api_key.create')
 
-        data["created_at"] = utc_now()
+        verrors = ValidationErrors()
+        self._validate("api_key_create", data, verrors)
+        user = self.middleware.call_sync('user.query', [
+            ['username', '=', data['username']]
+        ])
+        if not user:
+            verrors.add('api_key_create', 'User does not exist.')
+
+        if user and not user[0]['roles']:
+            verrors.add('api_key_create', 'User lacks privilege role membership.')
```

Comment:
I wanted it logically separated from validation errors processing because we're shifting authorization check from being handled before method call to internally in method. This makes it harder for a helpful edit to accidentally bypass authorization checks.